src/instamatic/processing/PETS_input_factory.py

```diff
@@ -0,0 +1,195 @@
+
+    @property
+    def list(self) -> list[str]:
+        return self.table.index.values.tolist()
+
+    def find(self, text: str) -> set[str]:
+        firsts = {t.strip().split()[0].lower() for t in text.splitlines() if t.strip()}
+        return firsts.intersection(self.list)
+
+
+with resources.files('instamatic.processing').joinpath('PETS_input_keywords.csv').open() as f:
+    pets_keywords = PetsKeywords.from_file(f)  # noqa: default keywords read at import
+
+
+@dataclass
+class PetsInputElement:
+    """Store metadata for a single PETS input element."""
+
+    keywords: list[str]
+    values: list[Any]
+    string: Optional[str] = None
+
+    def __str__(self):
+        return self.string if self.string is not None else self.build_string()
+
+    @classmethod
+    def from_any(cls, keyword_or_text: str, values: Optional[list[Any]] = None) -> Self:
+        keywords = pets_keywords.find(keyword_or_text)
+        if len(keywords) == 1 and values is not None:
+            return cls([keywords.pop()], values)
+        return cls(keywords=list(keywords), values=[], string=keyword_or_text)
+
+    @classmethod
+    def list_from_text(cls, text: str) -> list[Self]:
+        """Split a text with many commands it into `list[PetsInputElement]`"""
+        lines = text.splitlines()
+        firsts = [ts.split()[0].lower() if (ts := t.strip()) else '' for t in lines]
+
+        def split2blocks(start: int) -> list[str]:
+            if start >= len(lines):
+                return []
+            if firsts[start] not in pets_keywords.list:
+                return [lines[start]] + split2blocks(start + 1)
+            if pets_keywords.table.at[firsts[start], 'end'] == 'false':
+                return [lines[start]] + split2blocks(start + 1)
+            try:
+                end = firsts.index('end' + firsts[start], start)
+            except ValueError:
+                if pets_keywords.table.at[firsts[start], 'end'] == 'optional':
+                    return [lines[start]] + split2blocks(start + 1)
+                end = len(lines)
+            return ['\n'.join(lines[start : end + 1])] + split2blocks(end + 1)
+
+        return [PetsInputElement.from_any(block) for block in split2blocks(0)]
+
+    def has_end(self):
+        return (
+            len(self.keywords) == 1
+            and self.keywords[0] in pets_keywords.list
+            and pets_keywords.table.at[self.keywords[0], 'end'] != 'false'
+        )
+
+    def build_string(self):
+        if len(self.keywords) > 1:
+            warn(f'Building `str(PetsInputElement)` with >1 keywords! {self.keywords}')
+        prefix = [self.keywords[0]]
+        delimiter = '\n' if (has_end := self.has_end()) else ' '
+        suffix = [f'end{self.keywords[0]}'] if has_end else []
+        return delimiter.join(str(s) for s in prefix + self.values + suffix)
+
+
+AnyPetsInputElement = Union[PetsInputElement, str]
+
+
+class PetsInputWarning(UserWarning):
+    pass
```

Comment:
Can you move these to the top of the file?

```suggestion
AnyPetsInputElement = Union[PetsInputElement, str]


class PetsInputWarning(UserWarning):
    pass
```

Reply:
Good idea for `PetsInputWarning` but can't do it for `AnyPetsInputElement` since it uses `PetsInputElement` declared immediately above it, i.e. it is already as high as possible, and only helps with typing within objects immediately below it.